initial draft

joeniquette opened the pull request:
please add comments in your review

joeniquette committed:
initial draft

joeniquette committed:
format fix

AGarcia-Clarity commented on platform/engineering/design-docs/2022-01-12-Elixir.md:
```diff
@@ -0,0 +1,100 @@
+  - Varies based on complexity of the current service and solution, however the scaffolding and examples of other applications
+
+### Alternatives
+If Elixir does not deliver on the expected advantages we can either pivot to Go or Ruby. 
```
It would helpful to see the unique advantages of Elixir OTP/Phoenix framework listed out in a section versus the other alternatives of (Ruby, Nodejs, and Golang) (*Completely optional)

joeniquette replied:
I added more content here but I do welcome anyone wanting to add more to compare phoenix to rails

AGarcia-Clarity commented on platform/engineering/design-docs/2022-01-12-Elixir.md:
```diff
@@ -0,0 +1,100 @@
+- Highly performant language that allows for potentially 100s of thousands of concurrent connections
+- Built on Erlang virtual machine to leverage established stable technology which also allows for “active changes”, “reduced impact of errors”, and non-blocking IO
+- Requiring each response to the service to have minimal latency, Elixir performed better in many studies compared to Ruby, Node.
+- Ruby's current buckling of high requests results in exponential increase in response time, Elixir has a flat latency response under higher concurrency loads. - - This is extremely important for the sign in service as we expect to receive inconsistent network latency and exponentially growing requests during “high va.gov traffic hours”.
```
Is the next paragraph supposed to be a list item 👀 ?

joeniquette committed:
add content based on feedback

We certainly can do a better breakdown of phoenix vs rails but its Elixir that give us the advantages we are looking for

bramleyjl commented on platform/engineering/design-docs/2022-01-12-Elixir.md:
```diff
@@ -0,0 +1,100 @@
+ 
+
+### Background
+  The sign in service is intended to be a new solution for what will eventually be the authentication and authorization flow for all VA.gov users. This service will have clear concurrency and performance needs that do not make it a good candidate for an implementation inside the vets-api monolith, but rather encourage a separation as its own separate service. Additionally, the organization itself ...
```
Do we want to mention that this is going to scale beyond vets-api (since we're already handling MHV's sign in through vets-api) as another reason for needing excellent scalability. We could also elaborate that this is the best example of clear service boundaries - we don't want vets-api handling authentication for a sign in request that has nothing to do with vets-api.

f1337 replied:
Yeah I think that's important context. *I* know it, but all readers of this doc may not.

joeniquette replied:
I added some language in for this, if you could help add more I'd appreciate it. Im going to merge this in a few hours, future changes should open another PR

f1337 reviewed (approved):
I specifically like the CQS (command-query separation) architecture. Even if for some reason we pivot away from Elxir in the design (I don't foresee that as necessary), I think we should keep the CQS architecture.

Re: Elixir & Erlang => I have no objections. The Ruby community in general is pretty open-to and interested-in Elixir. And I hypothesize that folks from Java backgrounds will find Elixir a little bit less of a learning curve than Ruby (whose Smalltalk-derived design comes with gotchas).

joeniquette committed:
update format add paragraph

bosawt reviewed (approved):
Looks good

f1337 commented:
Because this is the public record of the discussion:

Rails is a terrible fit for this service. And I am a vocal Rails fanboy.

Our realistic choices are: Elixir, Go, Rust, Node. Of those, Elixir and Node are the best matches for our existing skill sets. Rails folks are generally amiable to learning Elixir+Phoenix, our current crop of Rails engineers included (we polled everyone a couple months back in Slack). Back end Node is do different than front end React development, that I think Elixir is actually the better fit.

Decision: Accepted.
